rust/gui-client/src-tauri/src/client/gui.rs

```diff
@@ -816,10 +823,16 @@ async fn run_controller(
                 }
             },
             resolvers = dns_listener.notified() => {
-                let resolvers = resolvers?;
                 if controller.status.connlib_is_up() {
-                    tracing::debug!(?resolvers, "New DNS resolvers, calling `Session::set_dns`");
-                    controller.ipc_client.set_dns(resolvers).await?;
+                    let resolvers = resolvers?;
+                    let resolver_set = HashSet::from_iter(resolvers.iter().cloned());
+                    if resolver_set != last_resolvers_sent {
```

Comment:
Eventloops are notoriously hard to test and tricky to get right. My advice would be that we either move this logic into connlib or you move it to your resolver detection and don't call connlib if they are the same.

What is actually the component that consumes a lot of time here?

Reply:
Maybe we can get rid of this optimization since the `set_dns` on Windows is faster now that it skips the deactivate step.

The problem is, to get really useful benchmarks I need to go to my Windows laptop, make a release build (or I guess, check if unoptimized builds put out the same numbers), and then run it in the VM a few times, so it's very slow every time I make any change.

Maybe since the tunnel stuff is splitting off I could do this:
- Set up benchmarks in CI that specify maximum times for the slowest I/O operations that don't need a portal connection, e.g. initialize WinTun, `set_dns`, deactivate DNS, set routes, set IPs, etc.
- Also set up those benchmarks as a diagnostic subcommand of the IPC service
- Measure the times just once and then I can separately ask, "Can I make this I/O faster?" and "Could I safely not do this I/O at all?", the second one taking a lot less back-and-forth since it's deterministic.

Reply:
> The problem is, to get really useful benchmarks I need to go to my Windows laptop, make a release build (or I guess, check if unoptimized builds put out the same numbers), and then run it in the VM a few times, so it's very slow every time I make any change.

I hear you. I have the same problem with testing Android changes because I haven't managed to build the Android app locally so I need to push to CI every time I want a new Android build :sweat_smile: 

I guess the issue is (at least) two-fold:

1. How to figure out, which parts of the stack are slow.
2. Making sure we don't regress in the future.

The `firezone-headless-client` works on Windows right? Perhaps some improvement can be achieved by testing things with that headless client directly because it is more scriptable or at least you can run the same command over and over again that just starts up the tunnel.

That is how I develop on connlib primarily: Build `firezone-headless-client` and run it with a service-account token against staging. It is quick and easy to adjust `RUST_LOG` etc.

Regarding release builds: I somewhat doubt that this makes a difference for these changes. You should be fine with a debug build. Unless I am planning to run an `iperf`, I never make a release build of the client in development. You are also missing out on debug assertions if you do that.

Reply:
Rolled this back since it doesn't seem to make a difference anymore